// cst/lib/elements/elementIndex.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _RestProperty = require('./types/RestProperty');

var _RestProperty2 = _interopRequireDefault(_RestProperty);

var _AssignmentExpression = require('./types/AssignmentExpression');

var _AssignmentExpression2 = _interopRequireDefault(_AssignmentExpression);

var _AssignmentPattern = require('./types/AssignmentPattern');

var _AssignmentPattern2 = _interopRequireDefault(_AssignmentPattern);

var _ArrayExpression = require('./types/ArrayExpression');

var _ArrayExpression2 = _interopRequireDefault(_ArrayExpression);

var _ArrayPattern = require('./types/ArrayPattern');

var _ArrayPattern2 = _interopRequireDefault(_ArrayPattern);

var _ArrowFunctionExpression = require('./types/ArrowFunctionExpression');

var _ArrowFunctionExpression2 = _interopRequireDefault(_ArrowFunctionExpression);

var _AwaitExpression = require('./types/AwaitExpression');

var _AwaitExpression2 = _interopRequireDefault(_AwaitExpression);

var _BinaryExpression = require('./types/BinaryExpression');

var _BinaryExpression2 = _interopRequireDefault(_BinaryExpression);

var _BindExpression = require('./types/BindExpression');

var _BindExpression2 = _interopRequireDefault(_BindExpression);

var _BlockStatement = require('./types/BlockStatement');

var _BlockStatement2 = _interopRequireDefault(_BlockStatement);

var _BooleanLiteral = require('./types/BooleanLiteral');

var _BooleanLiteral2 = _interopRequireDefault(_BooleanLiteral);

var _BreakStatement = require('./types/BreakStatement');

var _BreakStatement2 = _interopRequireDefault(_BreakStatement);

var _CallExpression = require('./types/CallExpression');

var _CallExpression2 = _interopRequireDefault(_CallExpression);

var _CatchClause = require('./types/CatchClause');

var _CatchClause2 = _interopRequireDefault(_CatchClause);

var _ClassBody = require('./types/ClassBody');

var _ClassBody2 = _interopRequireDefault(_ClassBody);

var _ClassDeclaration = require('./types/ClassDeclaration');

var _ClassDeclaration2 = _interopRequireDefault(_ClassDeclaration);

var _ClassExpression = require('./types/ClassExpression');

var _ClassExpression2 = _interopRequireDefault(_ClassExpression);

var _ClassMethod = require('./types/ClassMethod');

var _ClassMethod2 = _interopRequireDefault(_ClassMethod);

var _ClassProperty = require('./types/ClassProperty');

var _ClassProperty2 = _interopRequireDefault(_ClassProperty);

var _ConditionalExpression = require('./types/ConditionalExpression');

var _ConditionalExpression2 = _interopRequireDefault(_ConditionalExpression);

var _ContinueStatement = require('./types/ContinueStatement');

var _ContinueStatement2 = _interopRequireDefault(_ContinueStatement);

var _DebuggerStatement = require('./types/DebuggerStatement');

var _DebuggerStatement2 = _interopRequireDefault(_DebuggerStatement);

var _Directive = require('./types/Directive');

var _Directive2 = _interopRequireDefault(_Directive);

var _DirectiveLiteral = require('./types/DirectiveLiteral');

var _DirectiveLiteral2 = _interopRequireDefault(_DirectiveLiteral);

var _DoWhileStatement = require('./types/DoWhileStatement');

var _DoWhileStatement2 = _interopRequireDefault(_DoWhileStatement);

var _EmptyStatement = require('./types/EmptyStatement');

var _EmptyStatement2 = _interopRequireDefault(_EmptyStatement);

var _ExportAllDeclaration = require('./types/ExportAllDeclaration');

var _ExportAllDeclaration2 = _interopRequireDefault(_ExportAllDeclaration);

var _ExportDefaultDeclaration = require('./types/ExportDefaultDeclaration');

var _ExportDefaultDeclaration2 = _interopRequireDefault(_ExportDefaultDeclaration);

var _ExportDefaultSpecifier = require('./types/ExportDefaultSpecifier');

var _ExportDefaultSpecifier2 = _interopRequireDefault(_ExportDefaultSpecifier);

var _ExportNamedDeclaration = require('./types/ExportNamedDeclaration');

var _ExportNamedDeclaration2 = _interopRequireDefault(_ExportNamedDeclaration);

var _ExportNamespaceSpecifier = require('./types/ExportNamespaceSpecifier');

var _ExportNamespaceSpecifier2 = _interopRequireDefault(_ExportNamespaceSpecifier);

var _ExportSpecifier = require('./types/ExportSpecifier');

var _ExportSpecifier2 = _interopRequireDefault(_ExportSpecifier);

var _ExpressionStatement = require('./types/ExpressionStatement');

var _ExpressionStatement2 = _interopRequireDefault(_ExpressionStatement);

var _ForInStatement = require('./types/ForInStatement');

var _ForInStatement2 = _interopRequireDefault(_ForInStatement);

var _ForOfStatement = require('./types/ForOfStatement');

var _ForOfStatement2 = _interopRequireDefault(_ForOfStatement);

var _ForStatement = require('./types/ForStatement');

var _ForStatement2 = _interopRequireDefault(_ForStatement);

var _FunctionDeclaration = require('./types/FunctionDeclaration');

var _FunctionDeclaration2 = _interopRequireDefault(_FunctionDeclaration);

var _FunctionExpression = require('./types/FunctionExpression');

var _FunctionExpression2 = _interopRequireDefault(_FunctionExpression);

var _Identifier = require('./types/Identifier');

var _Identifier2 = _interopRequireDefault(_Identifier);

var _IfStatement = require('./types/IfStatement');

var _IfStatement2 = _interopRequireDefault(_IfStatement);

var _ImportDeclaration = require('./types/ImportDeclaration');

var _ImportDeclaration2 = _interopRequireDefault(_ImportDeclaration);

var _ImportDefaultSpecifier = require('./types/ImportDefaultSpecifier');

var _ImportDefaultSpecifier2 = _interopRequireDefault(_ImportDefaultSpecifier);

var _ImportNamespaceSpecifier = require('./types/ImportNamespaceSpecifier');

var _ImportNamespaceSpecifier2 = _interopRequireDefault(_ImportNamespaceSpecifier);

var _ImportSpecifier = require('./types/ImportSpecifier');

var _ImportSpecifier2 = _interopRequireDefault(_ImportSpecifier);

var _JSXAttribute = require('./types/JSXAttribute');

var _JSXAttribute2 = _interopRequireDefault(_JSXAttribute);

var _JSXClosingElement = require('./types/JSXClosingElement');

var _JSXClosingElement2 = _interopRequireDefault(_JSXClosingElement);

var _JSXElement = require('./types/JSXElement');

var _JSXElement2 = _interopRequireDefault(_JSXElement);

var _JSXEmptyExpression = require('./types/JSXEmptyExpression');

var _JSXEmptyExpression2 = _interopRequireDefault(_JSXEmptyExpression);

var _JSXExpressionContainer = require('./types/JSXExpressionContainer');

var _JSXExpressionContainer2 = _interopRequireDefault(_JSXExpressionContainer);

var _JSXIdentifier = require('./types/JSXIdentifier');

var _JSXIdentifier2 = _interopRequireDefault(_JSXIdentifier);

var _JSXMemberExpression = require('./types/JSXMemberExpression');

var _JSXMemberExpression2 = _interopRequireDefault(_JSXMemberExpression);

var _JSXNamespacedName = require('./types/JSXNamespacedName');

var _JSXNamespacedName2 = _interopRequireDefault(_JSXNamespacedName);

var _JSXOpeningElement = require('./types/JSXOpeningElement');

var _JSXOpeningElement2 = _interopRequireDefault(_JSXOpeningElement);

var _JSXSpreadAttribute = require('./types/JSXSpreadAttribute');

var _JSXSpreadAttribute2 = _interopRequireDefault(_JSXSpreadAttribute);

var _JSXText = require('./types/JSXText');

var _JSXText2 = _interopRequireDefault(_JSXText);

var _LabeledStatement = require('./types/LabeledStatement');

var _LabeledStatement2 = _interopRequireDefault(_LabeledStatement);

var _LogicalExpression = require('./types/LogicalExpression');

var _LogicalExpression2 = _interopRequireDefault(_LogicalExpression);

var _MemberExpression = require('./types/MemberExpression');

var _MemberExpression2 = _interopRequireDefault(_MemberExpression);

var _MetaProperty = require('./types/MetaProperty');

var _MetaProperty2 = _interopRequireDefault(_MetaProperty);

var _NewExpression = require('./types/NewExpression');

var _NewExpression2 = _interopRequireDefault(_NewExpression);

var _NullLiteral = require('./types/NullLiteral');

var _NullLiteral2 = _interopRequireDefault(_NullLiteral);

var _NumericLiteral = require('./types/NumericLiteral');

var _NumericLiteral2 = _interopRequireDefault(_NumericLiteral);

var _ObjectExpression = require('./types/ObjectExpression');

var _ObjectExpression2 = _interopRequireDefault(_ObjectExpression);

var _ObjectMethod = require('./types/ObjectMethod');

var _ObjectMethod2 = _interopRequireDefault(_ObjectMethod);

var _ObjectPattern = require('./types/ObjectPattern');

var _ObjectPattern2 = _interopRequireDefault(_ObjectPattern);

var _ObjectProperty = require('./types/ObjectProperty');

var _ObjectProperty2 = _interopRequireDefault(_ObjectProperty);

var _Program = require('./types/Program');

var _Program2 = _interopRequireDefault(_Program);

var _RegExpLiteral = require('./types/RegExpLiteral');

var _RegExpLiteral2 = _interopRequireDefault(_RegExpLiteral);

var _RestElement = require('./types/RestElement');

var _RestElement2 = _interopRequireDefault(_RestElement);

var _ReturnStatement = require('./types/ReturnStatement');

var _ReturnStatement2 = _interopRequireDefault(_ReturnStatement);

var _SequenceExpression = require('./types/SequenceExpression');

var _SequenceExpression2 = _interopRequireDefault(_SequenceExpression);

var _SpreadElement = require('./types/SpreadElement');

var _SpreadElement2 = _interopRequireDefault(_SpreadElement);

var _SpreadProperty = require('./types/SpreadProperty');

var _SpreadProperty2 = _interopRequireDefault(_SpreadProperty);

var _StringLiteral = require('./types/StringLiteral');

var _StringLiteral2 = _interopRequireDefault(_StringLiteral);

var _Super = require('./types/Super');

var _Super2 = _interopRequireDefault(_Super);

var _SwitchCase = require('./types/SwitchCase');

var _SwitchCase2 = _interopRequireDefault(_SwitchCase);

var _SwitchStatement = require('./types/SwitchStatement');

var _SwitchStatement2 = _interopRequireDefault(_SwitchStatement);

var _TaggedTemplateExpression = require('./types/TaggedTemplateExpression');

var _TaggedTemplateExpression2 = _interopRequireDefault(_TaggedTemplateExpression);

var _TemplateElement = require('./types/TemplateElement');

var _TemplateElement2 = _interopRequireDefault(_TemplateElement);

var _TemplateLiteral = require('./types/TemplateLiteral');

var _TemplateLiteral2 = _interopRequireDefault(_TemplateLiteral);

var _ThisExpression = require('./types/ThisExpression');

var _ThisExpression2 = _interopRequireDefault(_ThisExpression);

var _ThrowStatement = require('./types/ThrowStatement');

var _ThrowStatement2 = _interopRequireDefault(_ThrowStatement);

var _TryStatement = require('./types/TryStatement');

var _TryStatement2 = _interopRequireDefault(_TryStatement);

var _UnaryExpression = require('./types/UnaryExpression');

var _UnaryExpression2 = _interopRequireDefault(_UnaryExpression);

var _UpdateExpression = require('./types/UpdateExpression');

var _UpdateExpression2 = _interopRequireDefault(_UpdateExpression);

var _VariableDeclaration = require('./types/VariableDeclaration');

var _VariableDeclaration2 = _interopRequireDefault(_VariableDeclaration);

var _VariableDeclarator = require('./types/VariableDeclarator');

var _VariableDeclarator2 = _interopRequireDefault(_VariableDeclarator);

var _WithStatement = require('./types/WithStatement');

var _WithStatement2 = _interopRequireDefault(_WithStatement);

var _WhileStatement = require('./types/WhileStatement');

var _WhileStatement2 = _interopRequireDefault(_WhileStatement);

var _YieldExpression = require('./types/YieldExpression');

var _YieldExpression2 = _interopRequireDefault(_YieldExpression);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
    RestProperty: _RestProperty2.default,

    AssignmentExpression: _AssignmentExpression2.default,
    AssignmentPattern: _AssignmentPattern2.default,
    ArrayExpression: _ArrayExpression2.default,
    ArrayPattern: _ArrayPattern2.default,
    ArrowFunctionExpression: _ArrowFunctionExpression2.default,
    AwaitExpression: _AwaitExpression2.default,
    BinaryExpression: _BinaryExpression2.default,
    BindExpression: _BindExpression2.default,
    BlockStatement: _BlockStatement2.default,
    BooleanLiteral: _BooleanLiteral2.default,
    BreakStatement: _BreakStatement2.default,
    CallExpression: _CallExpression2.default,
    CatchClause: _CatchClause2.default,
    ClassBody: _ClassBody2.default,
    ClassDeclaration: _ClassDeclaration2.default,
    ClassExpression: _ClassExpression2.default,
    ClassMethod: _ClassMethod2.default,
    ClassProperty: _ClassProperty2.default,
    ConditionalExpression: _ConditionalExpression2.default,
    ContinueStatement: _ContinueStatement2.default,
    DebuggerStatement: _DebuggerStatement2.default,
    Directive: _Directive2.default,
    DirectiveLiteral: _DirectiveLiteral2.default,
    DoWhileStatement: _DoWhileStatement2.default,
    EmptyStatement: _EmptyStatement2.default,
    ExportAllDeclaration: _ExportAllDeclaration2.default,
    ExportDefaultDeclaration: _ExportDefaultDeclaration2.default,
    ExportDefaultSpecifier: _ExportDefaultSpecifier2.default,
    ExportNamedDeclaration: _ExportNamedDeclaration2.default,
    ExportNamespaceSpecifier: _ExportNamespaceSpecifier2.default,
    ExportSpecifier: _ExportSpecifier2.default,
    ExpressionStatement: _ExpressionStatement2.default,
    ForInStatement: _ForInStatement2.default,
    ForOfStatement: _ForOfStatement2.default,
    ForStatement: _ForStatement2.default,
    FunctionDeclaration: _FunctionDeclaration2.default,
    FunctionExpression: _FunctionExpression2.default,
    Identifier: _Identifier2.default,
    IfStatement: _IfStatement2.default,
    ImportDeclaration: _ImportDeclaration2.default,
    ImportDefaultSpecifier: _ImportDefaultSpecifier2.default,
    ImportNamespaceSpecifier: _ImportNamespaceSpecifier2.default,
    ImportSpecifier: _ImportSpecifier2.default,
    JSXAttribute: _JSXAttribute2.default,
    JSXClosingElement: _JSXClosingElement2.default,
    JSXElement: _JSXElement2.default,
    JSXEmptyExpression: _JSXEmptyExpression2.default,
    JSXExpressionContainer: _JSXExpressionContainer2.default,
    JSXIdentifier: _JSXIdentifier2.default,
    JSXMemberExpression: _JSXMemberExpression2.default,
    JSXNamespacedName: _JSXNamespacedName2.default,
    JSXOpeningElement: _JSXOpeningElement2.default,
    JSXSpreadAttribute: _JSXSpreadAttribute2.default,
    JSXText: _JSXText2.default,
    LabeledStatement: _LabeledStatement2.default,
    LogicalExpression: _LogicalExpression2.default,
    MemberExpression: _MemberExpression2.default,
    MetaProperty: _MetaProperty2.default,
    NewExpression: _NewExpression2.default,
    NullLiteral: _NullLiteral2.default,
    NumericLiteral: _NumericLiteral2.default,
    ObjectExpression: _ObjectExpression2.default,
    ObjectMethod: _ObjectMethod2.default,
    ObjectPattern: _ObjectPattern2.default,
    ObjectProperty: _ObjectProperty2.default,
    Program: _Program2.default,
    RegExpLiteral: _RegExpLiteral2.default,
    RestElement: _RestElement2.default,
    ReturnStatement: _ReturnStatement2.default,
    SequenceExpression: _SequenceExpression2.default,
    SpreadElement: _SpreadElement2.default,
    SpreadProperty: _SpreadProperty2.default,
    StringLiteral: _StringLiteral2.default,
    Super: _Super2.default,
    SwitchCase: _SwitchCase2.default,
    SwitchStatement: _SwitchStatement2.default,
    TaggedTemplateExpression: _TaggedTemplateExpression2.default,
    TemplateElement: _TemplateElement2.default,
    TemplateLiteral: _TemplateLiteral2.default,
    ThisExpression: _ThisExpression2.default,
    ThrowStatement: _ThrowStatement2.default,
    TryStatement: _TryStatement2.default,
    UnaryExpression: _UnaryExpression2.default,
    UpdateExpression: _UpdateExpression2.default,
    VariableDeclaration: _VariableDeclaration2.default,
    VariableDeclarator: _VariableDeclarator2.default,
    WithStatement: _WithStatement2.default,
    WhileStatement: _WhileStatement2.default,
    YieldExpression: _YieldExpression2.default
};
