// cst/lib/elements/types/ClassMethod.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _Node2 = require('../Node');

var _Node3 = _interopRequireDefault(_Node2);

var _getFunctionParams = require('./utils/getFunctionParams');

var _getFunctionParams2 = _interopRequireDefault(_getFunctionParams);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var getterAndSetter = {
    get: true,
    set: true
};

var ClassMethod = function (_Node) {
    (0, _inherits3.default)(ClassMethod, _Node);

    function ClassMethod(childNodes) {
        (0, _classCallCheck3.default)(this, ClassMethod);

        var _this = (0, _possibleConstructorReturn3.default)(this, (ClassMethod.__proto__ || (0, _getPrototypeOf2.default)(ClassMethod)).call(this, 'ClassMethod', childNodes));

        _this.isFunction = true;
        return _this;
    }

    (0, _createClass3.default)(ClassMethod, [{
        key: '_acceptChildren',
        value: function _acceptChildren(children) {
            var key = void 0;
            var computed = void 0;
            var kind = void 0;
            var staticMember = false;
            var generator = false;
            var async = false;
            var params = void 0;

            if (children.isToken('Identifier', 'static')) {
                staticMember = true;
                children.passToken();
                children.skipNonCode();
            }

            if (children.isToken('Identifier', 'async')) {
                async = true;
                children.passToken();
                children.skipNonCode();
            }

            if (children.isToken('Identifier', getterAndSetter)) {
                kind = children.currentElement.value;
                children.passToken();
                children.skipNonCode();
            } else {
                kind = 'method';
                if (children.isToken('Punctuator', '*')) {
                    generator = true;
                    children.passToken();
                    children.skipNonCode();
                }
            }

            if (children.isNode('Identifier')) {
                computed = false;
                key = children.passNode();
                if (kind === 'method' && key.type === 'Identifier' && key.name === 'constructor') {
                    kind = 'constructor';
                }
                children.skipNonCode();
            } else {
                computed = true;
                children.passToken('Punctuator', '[');
                children.skipNonCode();
                key = children.passExpression();
                children.skipNonCode();
                children.passToken('Punctuator', ']');
                children.skipNonCode();
            }

            if (children.isToken('Punctuator', '*')) {
                generator = true;
                children.moveNext();
                children.skipNonCode();
            }

            if (children.isNode('Identifier')) {
                children.passNode();
                children.skipNonCode();
            }

            params = (0, _getFunctionParams2.default)(children);
            children.skipNonCode();

            var body = children.passStatement();

            children.assertEnd();

            this.params = params;
            this.async = async;
            this.body = body;
            this.generator = generator;
            this.kind = kind;
            this.key = key;
            this.computed = computed;
            this.static = staticMember;
        }
    }]);
    return ClassMethod;
}(_Node3.default);

exports.default = ClassMethod;
