// cst/lib/elements/types/RegExpLiteral.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _Expression2 = require('../Expression');

var _Expression3 = _interopRequireDefault(_Expression2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var RegExpLiteral = function (_Expression) {
    (0, _inherits3.default)(RegExpLiteral, _Expression);

    function RegExpLiteral(childNodes) {
        (0, _classCallCheck3.default)(this, RegExpLiteral);
        return (0, _possibleConstructorReturn3.default)(this, (RegExpLiteral.__proto__ || (0, _getPrototypeOf2.default)(RegExpLiteral)).call(this, 'RegExpLiteral', childNodes));
    }

    (0, _createClass3.default)(RegExpLiteral, [{
        key: '_acceptChildren',
        value: function _acceptChildren(children) {
            children.assertToken('RegularExpression');
            var pattern = children.currentElement.value.pattern;
            var flags = children.currentElement.value.flags;
            var extra = {
                rawValue: undefined,
                raw: children.currentElement.getSourceCode()
            };

            children.moveNext();
            children.assertEnd();

            this.extra = extra;
            this.pattern = pattern;
            this.flags = flags;
        }
    }]);
    return RegExpLiteral;
}(_Expression3.default);

exports.default = RegExpLiteral;
